merino-suggest/src/lib.rs

```diff
@@ -291,6 +291,31 @@ pub trait SuggestionProvider: Send + Sync {
     fn is_null(&self) -> bool {
         false
     }
+
+    /// Generate a set of cache inputs for a given query specific to this
+    /// provider. Any property of the query that affects how suggestions are
+    /// generated should be included.
+    ///
+    /// By default, all properties of the query are used, but providers should
+    /// narrow this to a smaller scope.
+    fn cache_inputs(&self, req: &SuggestionRequest, hasher: &mut blake3::Hasher) {
+        hasher.update(req.query.as_bytes());
```

Comment:
The blake3 docs mention a performance concern if the input buffer is less than 16KiB. Not sure we can do anything about it right now, but I was wondering if you had checked it.

Reply:
I assume you're talking about the docs for `Hasher::update`, here? https://docs.rs/blake3/1.1.0/blake3/struct.Hasher.html#method.update

I read that more as 8KiB and 16KiB buffers get faster since it can use wider SIMD instructions. I wouldn't take that to mean there is a performance _problem_ for shorter inputs. I don't think it would be faster to artificially pad this out to 16KiB just to get into this fast case (though I'd be happy to see benchmarks contrary). I think the docs there are warning against naively using `std::io::copy` for hashing large inputs, since `std::io::copy` has a limited buffer size.

When we chose Blake3 for this hash [in this ADR](https://mozilla-services.github.io/merino/adrs/adr_0002_hashing.html#benchmarking-results-for-hashing-128-byte-values), I benchmarked all the candidates using 128 byte values, and even in that case Blake3 was the fastest network-safe hash that we used.

Reply:
Got it, thanks.